src/models.py

```diff
@@ -0,0 +1,135 @@
+        return nn.ReLU(inplace=True)
+    elif act_type == "gelu":
+        return nn.GELU()
+    elif act_type == "elu":
+        return nn.ELU()
+    # add other cases
+
+    else:
+        raise ValueError(f"Invalid activation layer type: {act_type}")
+
+
+def get_norm_layer(norm_type: str, hidden_features: int):
+    """
+    Returns a normalization layer based on the passed in type
+    """
+    if norm_type == "batch":
+        return nn.BatchNorm1d(hidden_features)
+    elif norm_type == "layer":
+        return nn.LayerNorm(hidden_features)
+    # add other cases
+
+    else:
+        raise ValueError(f"Invalid normalization layer type: {norm_type}")
+
+
+class DeeperGCN(nn.Module):
+    def __init__(
+        self,
+        input_features: int,
+        output_features: int,
+        hidden_features: int,
+        conv_type: str = "GEN",
+        act: str = "relu",
+        norm: str = "batch",
+        num_layers: int = 2,
+        use_cluster_pooling: bool = False,
+        readout: str = "mean",
+        dropout: float = 0.0,
+        edge_dim: Optional[int] = None,
+    ):
+        super().__init__()
+
+        self.readout = readout
+        self.use_cluster_pooling = use_cluster_pooling
+        self.dropout = dropout
+
+        # self.linears = nn.ModuleList()
+        # self.linears.append(nn.Linear(input_features, hidden_features))
+        # self.linears.append(nn.Linear(hidden_features, output_features))
+        self.fc_in = nn.Linear(input_features, hidden_features)
+        self.fc_out = nn.Linear(hidden_features, output_features)
+        self.out_act = get_act_layer(act)
+
+        self.layers = nn.ModuleList()
+        for _ in range(num_layers):
+            self.layers.append(
+                gnn.DeepGCNLayer(
+                    conv=get_conv_layer(conv_type, hidden_features, edge_dim=edge_dim),
+                    act=get_act_layer(act),
+                    norm=get_norm_layer(norm, hidden_features),
+                    block="res+",
+                    dropout=dropout,
+                )
+            )
+
+    def forward(self, data: Data) -> torch.Tensor:
+        # print(f"data: {data}")
+        # data.x = self.linears[0](data.x)
+        data.x = self.fc_in(data.x)
+
+        for layer in self.layers:
+            data.x = layer(data.x, data.edge_index, data.edge_attr)
+
+        x, batch = data.x, data.batch
+        if self.use_cluster_pooling:
+            data.edge_attr = None
+            weight = normalized_cut_2d(data.edge_index, data.pos)
+            cluster = gnn.graclus(data.edge_index, weight, data.x.size(0))
+            x, batch = gnn.max_pool_x(cluster, x, batch)
+
+        if self.readout == "mean":
+            x = gnn.global_mean_pool(x, batch)
+        elif self.readout == "max":
+            x = gnn.global_max_pool(x, batch)
+        elif self.readout == "add":
+            x = gnn.global_add_pool(x, batch)
+
+        F.dropout(x, p=self.dropout, training=self.training)
```

Comment:
Should the output be assigned to x? `x = F.dropout(x, ...)`

Reply:
Good eye! Fixed